### PR TITLE
Add Spatial/CanvasItem export property "inherits_transform"

### PR DESCRIPTION
**Pull request by willnationsdev**
The toplevel stuff is the flag that determines whether the Spatial should orient itself independently of its parent (so it doesn't "inherit" the parent's transform). I was thinking it should go in the Transform or Matrix area, but those seem to be dedicated exclusively to the vector values. It'd be much more convenient if we could just set this boolean from the Inspector rather than using a script or hacking it with a Node in between the two Spatials.

If anyone has suggestions for a different GROUP name or a different place to put this, let me know.
Let me know what you think.

### CODE REVIEW

**Comment by KellyThomas** [comment]
Can we do something similar to CanvasItem?

**Comment by willnationsdev** [comment]
@KellyThomas Oh, yeah, was looking in Node2D and didn't see it, but I guess CanvasItem does have it. I'll go ahead and add that in.

**Comment by willnationsdev** [comment]
Oh wait, looks like there's a comment from @bojidar-bg about not including it in the exports. Wonder what he thinks about it.

**Comment by bojidar-bg** [comment]
@willnationsdev I have no idea what I thought of then, but I think this comment was left by @reduz a few millenia before me.

Anyway, I think it would be better if it has some other name so that it doesn't conflict with existing scripts ("inherit_transform"?), and be visible in editor since it would be useful there.

**Comment by willnationsdev** [comment]
The name on the git blame is Bojidar iirc, so I figured it was you.

I'm all for changing the name to something clearer (figured it was just convention to keep the name the same-ish). I already set it up to be exposed to the Inspector - that was kinda the point of this PR.

**Comment by willnationsdev** [comment]
Oh wait, should probably change this so that I actually create "inherit_transform" setter/getters formally. Because their boolean value would actually be the opposite of "is_toplevel". lol

File: scene/2d/canvas_item.cpp [hunk 1]
```diff
@@ -1088,8 +1091,9 @@ void CanvasItem::_bind_methods() {
 	ADD_GROUP("Material", "");
 	ADD_PROPERTYNZ(PropertyInfo(Variant::OBJECT, "material", PROPERTY_HINT_RESOURCE_TYPE, "ShaderMaterial,CanvasItemMaterial"), "set_material", "get_material");
 	ADD_PROPERTYNZ(PropertyInfo(Variant::BOOL, "use_parent_material"), "set_use_parent_material", "get_use_parent_material");
+	ADD_GROUP("Context", "");
+	ADD_PROPERTYNZ(PropertyInfo(Variant::BOOL, "inherits_transform", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR), "set_inherits_transform", "get_inherits_transform");
```
Comment (bojidar-bg) [inline scene/2d/canvas_item.cpp]: Remove the `, PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR` part, since the defaults are better than this. Currently, this won't save the flag in the scene.

Also, it should be `ADD_PROPERTYNO`, so it saves a false value, but not a true value. (NZ is not-zero, NO is not-one)

File: scene/3d/spatial.cpp [hunk 2]
```diff
@@ -804,6 +807,8 @@ void Spatial::_bind_methods() {
 	ADD_GROUP("Visibility", "");
 	ADD_PROPERTYNO(PropertyInfo(Variant::BOOL, "visible"), "set_visible", "is_visible");
 	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "gizmo", PROPERTY_HINT_RESOURCE_TYPE, "SpatialGizmo", 0), "set_gizmo", "get_gizmo");
+	ADD_GROUP("Context", "");
+	ADD_PROPERTYNZ(PropertyInfo(Variant::BOOL, "inherits_transform", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR), "set_inherits_transform", "get_inherits_transform");
```
Comment (bojidar-bg) [inline scene/3d/spatial.cpp]: Same here.

**Review by bojidar-bg — CHANGES_REQUESTED** [review]

File: scene/2d/canvas_item.cpp (hunk 1 above)
Comment (willnationsdev) [reply scene/2d/canvas_item.cpp]: Ah, that makes sense. I made the change in both places.

**Comment by willnationsdev** [comment]
Not sure what that Travis error is about. Someone will need to restart that build I guess?

**Review by bojidar-bg — APPROVED** [review]
Might be nice if the old methods are marked as deprecated.

**Comment by willnationsdev** [comment]
@bojidar-bg just add a comment? Can do. In that case then, I'd move the methods to make them public rather than protected.

**Comment by bojidar-bg** [comment]
No, similar to `StaticBody2D::set_friction`:
https://github.com/godotengine/godot/blob/c02fb271fd54ecda331968d2269af79a6007dcb1/scene/2d/physics_body_2d.cpp#L190-L194

**Comment by reduz** [comment]
No, this function is more confusing because it makes it sound like inheriting transform is the exception, not the rule (which is the case). If you want to rename toplevel to something else its fine, but it must not invert the meaning. Also, if this is changed it will have to be for 4.0, as this is a pretty big breakage even it backwards compatibility is kept.

**Comment by reduz** [comment]
Also, I dont think the editor is designed to work well with this change (exposed property), so editor must be updated too.

**Comment by reduz** [comment]
In my opinion, this should not be exposed to editor since it makes no sense there. Same as process and other properties. If you need this, its going to be from code, so it can be enabled there.

**Comment by willnationsdev** [comment]
As per discussion in IRC, closing this PR.

TL;DR: The typical, intended workflow of using scenes is for them to work within the editor as if they are placed relatively. Forcing an absolute position is more often a runtime consideration, so setting the flag *at* runtime via script vs. at editor-time via Inspector is cleaner. If a user wants the even-more-rare workflow of positioning it absolutely in-editor relative to its parent, then they can use a Node buffer.

    <reduz> Juan Linietsky willnationsdev: again, it prevents you from instancing, so its kind of pointless
    12:03 PM you will edit the scene in the editor, then if you want to instance it, position and edit that scene some somewhere else, it wont work
    12:04 PM all the times you instance the scene will put the node in the same place
    12:04 PM so its kind of useless as a feature in general
    12:04 PM <willnationsdev> reduz: Ah, I gotcha now. It creates an exceptional behavior as the new default when dealing with that scene alone.
    12:05 PM That makes sense.
    12:05 PM <reduz> Juan Linietsky far most of the time, you will want to position it properly, instance it properly and then once it runs make it not inherit transform
    12:05 PM so it at least retains position
    12:05 PM this is why the property is not exposed, its common sense

**Comment by aaronfranke** [comment]
For future reference, if you want to make a Spatial or CanvasItem not inherit the transform, you can simply make it a child of a `Node`.
